# Chapter 2: Intron Peaks over heat response genes

# Load libraries
library(GenomicRanges)
library(ChIPseeker)
library(clusterProfiler)
library(GenomicFeatures)
library(DESeq2)
library(ggplot2)
library(ggVennDiagram)
library(ggdist)
library(gghalves)
library(ggpubr)
library(patchwork)
library(introdataviz)


# Load in data sets
# Transcripts
gff.file <- "~/Lab Notebook/Chapter2/Data_analysis/ATAC/Amil_ref_gen/Amil_v2.01/Amil.all.maker.noseq.gff"
txdb <- makeTxDbFromGFF(gff.file, format = "gff3") 
a.millepora.txs <- transcripts(txdb)

# Load in narrowPeaks files generated by MACS2 for significant peaks
narrowpeaks_paths <- c("~/Lab Notebook/Chapter2/Data_analysis/ATAC/Peak_files/AMIL_GENO_01_S1_peaks.narrowPeak",
                       "~/Lab Notebook/Chapter2/Data_analysis/ATAC/Peak_files/AMIL_GENO_02_S2_peaks.narrowPeak",
                       "~/Lab Notebook/Chapter2/Data_analysis/ATAC/Peak_files/AMIL_GENO_03_S3_peaks.narrowPeak",
                       "~/Lab Notebook/Chapter2/Data_analysis/ATAC/Peak_files/AMIL_GENO_04_S4_peaks.narrowPeak")

narrowpeaks.gr.list <- list()

for (file_path in narrowpeaks_paths) {
  narrow.peaks <- readPeakFile(file_path)
  narrowpeaks.gr.list[[file_path]] <- narrow.peaks
}

narrowpeaks.colnames <- c("name",
                          "score",
                          "V6",
                          "signalValue", #overall enrichment in the region
                          "pValue", # -log10
                          "qValue", # -log10 FDR
                          "peak") # point-source of the peak

colnames(mcols(narrowpeaks.gr.list[[1]])) <- narrowpeaks.colnames
colnames(mcols(narrowpeaks.gr.list[[2]])) <- narrowpeaks.colnames
colnames(mcols(narrowpeaks.gr.list[[3]])) <- narrowpeaks.colnames
colnames(mcols(narrowpeaks.gr.list[[4]])) <- narrowpeaks.colnames
head(narrowpeaks.gr.list[[1]])
head(narrowpeaks.gr.list[[2]])
head(narrowpeaks.gr.list[[3]])
head(narrowpeaks.gr.list[[4]])

# Peak Annotation
peakAnno.1 <- annotatePeak(narrowpeaks.gr.list[[1]], tssRegion=c(-3000, 3000),
                           TxDb=txdb)

peakAnno.2 <- annotatePeak(narrowpeaks.gr.list[[2]], tssRegion=c(-3000, 3000),
                           TxDb=txdb)

peakAnno.3 <- annotatePeak(narrowpeaks.gr.list[[3]], tssRegion=c(-3000, 3000),
                           TxDb=txdb)

peakAnno.4 <- annotatePeak(narrowpeaks.gr.list[[4]], tssRegion=c(-3000, 3000),
                           TxDb=txdb)


# Annotated peaks data frames
peaks.df.1 <- as.data.frame(peakAnno.1@anno@elementMetadata@listData)
genomic.annotation.detail.df.1 <- peakAnno.1@detailGenomicAnnotation
peaks.df.1 <- cbind(peaks.df.1, genomic.annotation.detail.df.1)

peaks.df.2 <- as.data.frame(peakAnno.2@anno@elementMetadata@listData)
genomic.annotation.detail.df.2 <- peakAnno.2@detailGenomicAnnotation
peaks.df.2 <- cbind(peaks.df.2, genomic.annotation.detail.df.2)

peaks.df.3 <- as.data.frame(peakAnno.3@anno@elementMetadata@listData)
genomic.annotation.detail.df.3 <- peakAnno.3@detailGenomicAnnotation
peaks.df.3 <- cbind(peaks.df.3, genomic.annotation.detail.df.3)

peaks.df.4 <- as.data.frame(peakAnno.4@anno@elementMetadata@listData)
genomic.annotation.detail.df.4 <- peakAnno.4@detailGenomicAnnotation
peaks.df.4 <- cbind(peaks.df.4, genomic.annotation.detail.df.4)

# All introns

geno.1.all.intron <- filter(peaks.df.1, Intron == TRUE) 

geno.2.all.intron <- filter(peaks.df.2, Intron == TRUE)  

geno.3.all.intron <- filter(peaks.df.3, Intron == TRUE) 

geno.4.all.intron <- filter(peaks.df.4, Intron == TRUE)


# Moving forward with looking at all introns
# Filter introns for unique gene names and adjusted pValue

pval.threshold <- -log10(0.00005)

geno.1.all.intron <- filter(geno.1.all.intron, pValue >= pval.threshold) 
genes.geno.1<- unique(data.frame(gene = geno.1.all.intron$geneId))
genes.geno.1$geno<- as.factor("1")
dim(genes.geno.1) #1842


geno.2.all.intron <- filter(geno.2.all.intron, pValue >= pval.threshold)
genes.geno.2 <- unique(data.frame(gene = geno.2.all.intron$geneId))
genes.geno.2$geno<- as.factor("2")
dim(genes.geno.2) #9888

geno.3.all.intron <- filter(geno.3.all.intron, pValue >= pval.threshold) 
genes.geno.3 <- unique(data.frame(gene = geno.3.all.intron$geneId))
genes.geno.3$geno<- as.factor("3")
dim(genes.geno.3) #1550

geno.4.all.intron <- filter(geno.4.all.intron, pValue >= pval.threshold) 
genes.geno.4 <- unique(data.frame(gene = geno.4.all.intron$geneId))
genes.geno.4$geno<- as.factor("4")
dim(genes.geno.4) #3567

# Visualize density of peaks over TSS:

geno1_density <- ggplot(geno.1.all.intron, aes(x = distanceToTSS)) +
  geom_density(fill = "skyblue", alpha = 0.5) +
  labs(title = "Geno 1: Distance to TSS",
       x = "Distance to TSS",
       y = "Density") +
  theme_pubr()

geno2_density <- ggplot(geno.2.all.intron, aes(x = distanceToTSS)) +
  geom_density(fill = "skyblue", alpha = 0.5) +
  labs(title = "Geno 2: Distance to TSS",
       x = "Distance to TSS",
       y = "Density") +
  theme_pubr()

geno3_density <- ggplot(geno.3.all.intron, aes(x = distanceToTSS)) +
  geom_density(fill = "skyblue", alpha = 0.5) +
  labs(title = "Geno 3: Distance to TSS",
       x = "Distance to TSS",
       y = "Density") +
  theme_pubr()

geno4_density <- ggplot(geno.4.all.intron, aes(x = distanceToTSS)) +
  geom_density(fill = "skyblue", alpha = 0.5) +
  labs(title = "Geno 4: Distance to TSS",
       x = "Distance to TSS",
       y = "Density") +
  theme_pubr()

(geno1_density + geno2_density) / (geno3_density + geno4_density)

# Visualize overlap
x <- list(
  Geno_1 = genes.geno.1$gene,
  Geno_2 = genes.geno.2$gene,
  Geno_3 = genes.geno.3$gene,
  Geno_4 = genes.geno.4$gene
)
ggVennDiagram(x)


# Isolate genes that are in peaks in at least 2 samples (there should be 3753 based on Venn)
# Looking at peaks in 2/4 individuals

geneID_geno1 <- unique(as.data.frame(geno.1.all.intron$geneId))
colnames(geneID_geno1) <- "geneID"
geneID_geno2 <- unique(as.data.frame(geno.2.all.intron$geneId))
colnames(geneID_geno2) <- "geneID"
geneID_geno3 <- unique(as.data.frame(geno.3.all.intron$geneId))
colnames(geneID_geno3) <- "geneID"
geneID_geno4 <- unique(as.data.frame(geno.4.all.intron$geneId))
colnames(geneID_geno4) <- "geneID"

geneID_1_2 <- dplyr::inner_join(geneID_geno1, geneID_geno2, by = "geneID")
geneID_1_3 <- dplyr::inner_join(geneID_geno1, geneID_geno3, by = "geneID")
geneID_1_4 <- dplyr::inner_join(geneID_geno1, geneID_geno4, by = "geneID")

geneID_2_3 <- dplyr::inner_join(geneID_geno2, geneID_geno3, by = "geneID")
geneID_2_4 <- dplyr::inner_join(geneID_geno2, geneID_geno4, by = "geneID")

geneID_3_4 <- dplyr::inner_join(geneID_geno3, geneID_geno4, by = "geneID")

geneID_2_outta_4 <- rbind(geneID_1_2, 
                          geneID_1_3, 
                          geneID_1_4, 
                          geneID_2_3, 
                          geneID_2_4,
                          geneID_3_4)

geneID_2_outta_4 <- unique(geneID_2_outta_4$geneID) #3753

geneID_2_outta_4.df <- as.data.frame(geneID_2_outta_4)
colnames(geneID_2_outta_4.df) <- "gene"
geneID_2_outta_4.df$num.genos <- "two"

## Isolate genes that are in peaks in at least 1 sample
geneID_1_outta_4 <- rbind(geneID_geno1, 
                          geneID_geno2, 
                          geneID_geno3, 
                          geneID_geno4)
geneID_1_outta_4 <- unique(geneID_1_outta_4$geneID)
geneID_1_outta_4.df <- as.data.frame(geneID_1_outta_4)
colnames(geneID_1_outta_4.df) <- "gene"
geneID_1_outta_4.df$num.genos <- "one"
## Isolate genes that are in peaks in at least 3 samples
geneID_1_2_3 <- dplyr::inner_join(geneID_1_2, geneID_geno3, by = "geneID")
geneID_1_2_4 <- dplyr::inner_join(geneID_1_2, geneID_geno4, by = "geneID")
geneID_1_3_4 <- dplyr::inner_join(geneID_1_3, geneID_geno4, by = "geneID")
geneID_2_3_4 <- dplyr::inner_join(geneID_2_3, geneID_geno4, by = "geneID")

geneID_3_outta_4 <- rbind(geneID_1_2_3,
                          geneID_1_2_4,
                          geneID_1_3_4,
                          geneID_2_3_4)

geneID_3_outta_4 <- unique(geneID_3_outta_4$geneID) # 843
geneID_3_outta_4.df <- as.data.frame(geneID_3_outta_4)
colnames(geneID_3_outta_4.df) <- "gene"
geneID_3_outta_4.df$num.genos <- "three"

## Isolate genes that are in peaks in all 4 samples
geneID_4_outta_4 <- dplyr::inner_join(geneID_1_2_3, geneID_geno4, by = "geneID")
geneID_4_outta_4 <- unique(geneID_4_outta_4$geneID) #235
geneID_4_outta_4.df <- as.data.frame(geneID_4_outta_4)
colnames(geneID_4_outta_4.df) <- "gene"
geneID_4_outta_4.df$num.genos <- "four"

# Open chromatin introns data frame
genes.with.open.chrm.introns <- rbind(geneID_1_outta_4.df,
                                        geneID_2_outta_4.df,
                                        geneID_3_outta_4.df,
                                        geneID_4_outta_4.df)

# Load in D0 Heat response genes
D0.HRGs <- read.csv('~/Lab Notebook/Chapter2/Data_analysis/TagSeq/Data_out/D0.heat.response.genes.csv', header = TRUE)

# ID genes in peaks in genes 
in.peaks.two.genos <- D0.HRGs$gene %in% geneID_2_outta_4 #  genes overlap
D0.HRGs$two.genos <- in.peaks.two.genos

# log10 transform the baseMean
D0.HRGs$log10bM <- log10(D0.HRGs$baseMean)

# Sort the absolute values for plotting
sorted.log10bM <- sort(D0.HRGs$log10bM)

# Create a Q-Q plot
qqnorm(sorted.log10bM)
qqline(sorted.log10bM )

# Shapiro-Wilk test for normality
shapiro.test(D0.HRGs$log10bM) # Reject the null that the data is normally distributed

# Log10 Base mean expression of genes with accessible introns
library(ggpubr)
rain_height <- .1

basemean.plot <- ggplot(D0.HRGs, aes(x = two.genos, y = log10bM, fill = two.genos )) +
  # clouds
  introdataviz::geom_flat_violin(trim=FALSE,
                                 position = position_nudge(x = rain_height+.05)) +
  # rain
  geom_point(aes(color = two.genos), size = 2, alpha = 0.4,  show.legend = FALSE, 
             position = position_jitter(width = rain_height, height = 0)) +
  # boxplots
  geom_boxplot(width = rain_height, show.legend = FALSE, 
               outlier.shape = NA,
               position = position_nudge(x = -rain_height*2)) +
  coord_flip() +
  scale_fill_manual(values = c("#f7b267",
                               "#f27059")) +
  scale_color_manual(values = c("#f7b267",
                                "#f27059")) + 
  
  theme_pubr() +
  stat_compare_means(method = "wilcox.test",
                     comparisons = list(c("TRUE", "FALSE")),
                     label = "p.format",
                     step.increase = 0.05) +
  theme(legend.position = "none") +
  labs(x = "Accessible intron", y = "Base Mean (log10)") +
  ggtitle("Base mean of HRGs\nof Accessible & Non-accessible introns\n(Day 0 Samples)")

# Of the DO hrgs, what is the absolute value of the log2fc for genes in peaks and
# genes not in peaks?

# Is the absolute value of the log2FC normally distributed

# Sort the absolute values for plotting
sorted.abs.log2FoldChange <- sort(abs(D0.HRGs$log2FoldChange))

# Create a Q-Q plot
qqnorm(sorted.abs.log2FoldChange)
qqline(sorted.abs.log2FoldChange)

# Shapiro-Wilk test for normality
shapiro.test(sorted.abs.log2FoldChange) # Reject the null that the data is normally distributed

l2fc.plot <- ggplot(D0.HRGs, aes(x = two.genos, y = abs(log2FoldChange), fill = two.genos )) +
  # clouds
  introdataviz::geom_flat_violin(trim=FALSE,
                                 position = position_nudge(x = rain_height+.05)) +
  # rain
  geom_point(aes(color = two.genos), size = 2, alpha = 0.4,  show.legend = FALSE, 
             position = position_jitter(width = rain_height, height = 0)) +
  # boxplots
  geom_boxplot(width = rain_height, show.legend = FALSE, 
               outlier.shape = NA,
               position = position_nudge(x = -rain_height*2)) +
  coord_flip() +
  scale_fill_manual(values = c("#f7b267",
                               "#f27059")) +
  scale_color_manual(values = c("#f7b267",
                                "#f27059")) + 
  
  theme_pubr() +
  stat_compare_means(method = "wilcox.test",
                     comparisons = list(c("TRUE", "FALSE")),
                     label = "p.format",
                     step.increase = 0.05) +
  theme(legend.position = "none") +
  labs(x = "Accessible intron", y = "Log2FC (absolute value)") +
  ggtitle("Log2FC of HRGs\nof Accessible & Non-accessible introns\n(Day 0 Samples)")


# overall expression variation by taking the standard deviation across all 
# normalized counts in all individuals and in controls.
## Expression data

file.dir <- "/Users/tillandsia/Lab Notebook/Chapter2/Data_analysis/TagSeq/Gene_count_files/"
my.files <- grep(".txt", list.files(file.dir), value=TRUE)
my.metadata <- read.table("/Users/tillandsia/Lab Notebook/Chapter2/Data_analysis/TagSeq/tag_seq_meta_data.txt", header = TRUE)

# Create sample table
sampleNames <- c("01_27_Y","01_28_R","01_40_Y","01_42_R","01_45_Y","01_46_Y",
                 "01_51_Y","01_52_Y","01_54_Y","01_58_Y","01_63_Y","01_68_Y",
                 "01_78_R","01_92_Y","02_29_Y","02_34_Y","02_35_R","02_47_Y",
                 "02_65_R","02_79_R","03_29_R","03_48_Y","03_62_R","03_72_R",
                 "03_90_Y","03_91_R","04_48_R","04_65_Y","04_70_Y","04_95_Y",
                 "05_43_R","05_46_R","05_50_Y","05_62_Y","05_82_Y","05_88_R",
                 "06_22_Y","06_32_R","06_59_R","06_64_R","06_67_R","06_72_Y",
                 "06_82_R","06_92_R","07_31_R","07_36_R","07_39_R","07_53_Y",
                 "07_64_Y","07_88_Y","08_19_R","08_30_R","08_33_Y","08_36_Y",
                 "08_45_R","08_50_R","08_85_Y","08_95_R")

my.sampleTable <- data.frame(sampleName = sampleNames, 
                             fileName = my.files, 
                             condition = my.metadata[,2:10])

# Convert variables to factors
factorVars <- c("condition.acclimation.temp", "condition.acclimation.class", "condition.acclimation.txt",
                "condition.stress.txt", "condition.stress.group", 
                "condition.stress.day", "condition.ext.date",
                "condition.ext.batch", "condition.genotype")


my.sampleTable[, factorVars] <- lapply(my.sampleTable[, factorVars], factor)
colnames(my.sampleTable)

# Check my.sampleTable
my.sampleTable

# Create DESeqDataSet Object
ddsHTSeq <- DESeqDataSetFromHTSeqCount(
  sampleTable = my.sampleTable,
  directory = file.dir,
  design = ~ condition.acclimation.txt +
    condition.stress.txt +
    condition.stress.day +
    condition.acclimation.txt:condition.stress.txt)

# Pre-filter
keep <- rowSums(counts(ddsHTSeq)) >= 10
ddsHTSeq <- ddsHTSeq[keep,]


ddsHTSeq$condition.acclimation.txt <- relevel(ddsHTSeq$condition.acclimation.txt, "control")
ddsHTSeq$condition.acclimation.class <- relevel(ddsHTSeq$condition.acclimation.class, "control")

# Extract normalized counts matrix
# Estimate size factors
ddsHTSeq <- estimateSizeFactors(ddsHTSeq)

# Estimate normalization factors
ddsHTSeq <- estimateDispersions(ddsHTSeq)
normalized.counts <- as.data.frame(counts(ddsHTSeq, normalized = TRUE))
normalized.counts$gene <- rownames(normalized.counts) 
normalized.counts$standard.deviation <- apply(normalized.counts[,1:58], 1, sd)

# Merge this with all genes in peaks (in at least 2 out of 4 samples)
# ID genes in peaks in genes 
in.peaks.two.genos.exp.var <- normalized.counts$gene %in% geneID_2_outta_4 #  genes overlap
normalized.counts$two.genos <- in.peaks.two.genos.exp.var
sum(normalized.counts$two.genos) # 2205


# Standard deviation of normalized gene counts genes with accessible introns
library(ggpubr)
rain_height <- .1

# There is an outlier here. Let's remove it and rerun the stats and plot
# ggplot(normalized.counts, aes(x = two.genos, y = standard.deviation, fill = two.genos )) +
#   # clouds
#   introdataviz::geom_flat_violin(trim=FALSE,
#                                  position = position_nudge(x = rain_height+.05)) +
#   # rain
#   geom_point(aes(color = two.genos), size = 2, alpha = 0.4,  show.legend = FALSE, 
#              position = position_jitter(width = rain_height, height = 0)) +
#   # boxplots
#   geom_boxplot(width = rain_height, show.legend = FALSE, 
#                outlier.shape = NA,
#                position = position_nudge(x = -rain_height*2)) +
#   coord_flip() +
#   scale_fill_manual(values = c("#909580",
#                                "#e7ad99")) +
#   scale_color_manual(values = c("#909580",
#                                 "#e7ad99")) + 
#   
#   theme_pubr() +
#   stat_compare_means(method = "wilcox.test",
#                      comparisons = list(c("TRUE", "FALSE")),
#                      label = "p.format",
#                      step.increase = 0.05) +
#   theme(legend.position = "none")

# Filtered normalized counts
filtered.normalized.counts <- normalized.counts[normalized.counts$standard.deviation <= 20000, ]


# Test for normality of log10(sd) normalized counts
# Sort the absolute values for plotting
sorted.log10.sd <- sort(log10(filtered.normalized.counts$standard.deviation))

# Create a Q-Q plot
qqnorm(sorted.log10.sd)
qqline(sorted.log10.sd)

# Kolmogorov-Smirnov test for normality (can be applied to dataset > 5000)
ks.test(sorted.log10.sd, "pnorm", mean = mean(sorted.log10.sd), sd = sd(sorted.log10.sd)) # reject the null that the data is normally distributed

exp.variation.plot <- ggplot(filtered.normalized.counts, aes(x = two.genos, y = log10(standard.deviation), fill = two.genos )) +
  # clouds
  introdataviz::geom_flat_violin(trim=FALSE,
                                 position = position_nudge(x = rain_height+.05)) +
  # rain
  geom_point(aes(color = two.genos), size = 2, alpha = 0.4,  show.legend = FALSE, 
             position = position_jitter(width = rain_height, height = 0)) +
  # boxplots
  geom_boxplot(width = rain_height, show.legend = FALSE, 
               outlier.shape = NA,
               position = position_nudge(x = -rain_height*2)) +
  coord_flip() +
  scale_fill_manual(values = c("#dec0f1",
                               "#aea3b0")) +
  scale_color_manual(values = c("#dec0f1",
                                "#aea3b0")) + 
  
  theme_pubr() +
  stat_compare_means(method = "wilcox.test",
                     comparisons = list(c("TRUE", "FALSE")),
                     label = "p.format",
                     step.increase = 0.05) +
  theme(legend.position = "none") +
  
  labs(x = "Open intron", y = "Standard deviation of normalized counts (log10)") +
  ggtitle("Overall expression variation\nof open & closed introns")

# Wilcox test of the normalized gene counts

# Extract standard deviation for each group
sd.true <- filtered.normalized.counts$standard.deviation[filtered.normalized.counts$two.genos == TRUE]
sd.false <- filtered.normalized.counts$standard.deviation[filtered.normalized.counts$two.genos == FALSE]

wilcox.test(sd.true, sd.false)



# Print all three
exp.variation.plot | (basemean.plot / l2fc.plot)

# Contingency table for the set of genes where I have peak and ge data for:

# peak data-total set of peaks (intron accessible in at least 1 out of four samples):
## see formation of x for filtering criteria
## pvalue filtered
union.peaks <- Reduce(union, x)
union.peaks <-as.data.frame(union.peaks)
colnames(union.peaks) <- "gene"
dim(union.peaks)
# Filter total.gene.set for genes that I also have peak data for
total.gene.set <- read.csv('~/Lab Notebook/Chapter2/Data_analysis/TagSeq/Data_out/total.set.gene.names.csv', header = TRUE)
total.gene.and.peak <- dplyr::inner_join(union.peaks, total.gene.set, by = "gene") #5872

# Filter total.gene.and.peak for peaks in at least 2 out 4 samples 
accessible.peaks <- dplyr::inner_join(total.gene.and.peak,geneID_2_outta_4.df, by = "gene") #1208

# Day 0 HRGs
D0.HRG.set <- as.data.frame(D0.HRGs$gene)
colnames(D0.HRG.set) <- "gene"

# Count the overlap (yes HRG/yes Peak)
yes.HRG.yes.Peak <- sum(D0.HRG.set$gene %in% union.peaks$gene)
yes.HRG.yes.Peak #390

# Set up contingency table
total.genes <- nrow(total.gene.and.peak)
HRG.count <- nrow(D0.HRG.set)
accessible.intron.genes.count <- nrow(accessible.peaks)
overlap.count <- yes.HRG.yes.Peak

# Create a 2x2 contingency table
contingency.table <- matrix(c(overlap.count, HRG.count - overlap.count,
                              accessible.intron.genes.count - overlap.count,
                              total.genes - HRG.count - accessible.intron.genes.count + overlap.count),
                            nrow = 2, ncol = 2, byrow = TRUE)

# Perform Fisher's exact test
result <- fisher.test(contingency.table)

# Extract p-value
p.value <- result$p.value
p.value

#base expression overall (all genes)
dds <- DESeq(ddsHTSeq)
hs.results <- lfcShrink(dds,
                        contrast = c("condition.stress.txt", "36", "24.5"),
                        type = "ashr")

heat.response.results <- hs.results[complete.cases(hs.results),]
heat.response.results$gene <- rownames(heat.response.results)

# Merge this with all genes in peaks (in at least 2 out of 4 samples)
# ID genes in peaks in genes 
in.peaks.two.genos.base.mean<- heat.response.results$gene %in% geneID_2_outta_4 #  genes overlap
heat.response.results$two.genos <- in.peaks.two.genos.base.mean
sum(heat.response.results$two.genos) # 1800

# Test the normality of the log10(baseMean) 
# log10 transform the baseMean
heat.response.results$log10bM <- log10(heat.response.results$baseMean)

# Sort the absolute values for plotting
sorted.log10bM.all.samples <- sort(heat.response.results$log10bM)

# Create a Q-Q plot
qqnorm(sorted.log10bM.all.samples)
qqline(sorted.log10bM.all.samples)

# Kolmogorov-Smirnov test for normality (can be applied to dataset > 5000)
ks.test(sorted.log10bM.all.samples, "pnorm", mean = mean(sorted.log10bM.all.samples), sd = sd(sorted.log10bM.all.samples)) # reject the null that the data is normally distributed


all.samples.basemean <- ggplot(heat.response.results, aes(x = two.genos, y = log10(baseMean), fill = two.genos )) +
  # clouds
  introdataviz::geom_flat_violin(trim=FALSE,
                                 position = position_nudge(x = rain_height+.05)) +
  # rain
  geom_point(aes(color = two.genos), size = 2, alpha = 0.4,  show.legend = FALSE, 
             position = position_jitter(width = rain_height, height = 0)) +
  # boxplots
  geom_boxplot(width = rain_height, show.legend = FALSE, 
               outlier.shape = NA,
               position = position_nudge(x = -rain_height*2)) +
  coord_flip() +
  scale_fill_manual(values = c("#dec0f1",
                               "#aea3b0")) +
  scale_color_manual(values = c("#dec0f1",
                                "#aea3b0")) + 
  
  theme_pubr() +
  stat_compare_means(method = "wilcox.test",
                     comparisons = list(c("TRUE", "FALSE")),
                     label = "p.format",
                     step.increase = 0.05) +
  theme(legend.position = "none") +
  
  labs(x = "Open intron", y = "Base Mean (log10)") +
  ggtitle("Overall base expression\nof open & closed introns")

# Wilcox test on untransformed values:
heat.response.results <- as.data.frame(heat.response.results)
# Extract baseMean values for each group
baseMean.true <- heat.response.results$baseMean[heat.response.results$two.genos == TRUE]
baseMean.false <- heat.response.results$baseMean[heat.response.results$two.genos == FALSE]

wilcox.test(baseMean.true, baseMean.false)

# Print all 4
(all.samples.basemean/exp.variation.plot) | (basemean.plot / l2fc.plot)


### Series of density plots by open regions
basemeans.open.chrom <- dplyr::full_join(heat.response.results, genes.with.open.chrm.introns, by = "gene") 
basemeans.open.chrom$num.genos[is.na(basemeans.open.chrom$num.genos)] <- "never"

num.genos.order <- c("never", "four",
                     "three", "two", "one")
basemeans.open.chrom$num.genos <- factor(basemeans.open.chrom$num.genos, levels = num.genos.order)
basemeans.open.chrom <- basemeans.open.chrom[is.finite(basemeans.open.chrom$baseMean), ]
all.samples.basemean.multi.genos <- ggplot(basemeans.open.chrom, aes(x = num.genos, y = log10(baseMean), fill = num.genos)) +
  # clouds
  introdataviz::geom_flat_violin(trim=TRUE,
                                 position = position_nudge(x = rain_height+.05)) +
  # rain
  geom_point(aes(color = num.genos), size = 2, alpha = 0.4,  show.legend = FALSE, 
             position = position_jitter(width = rain_height, height = 0)) +
  # boxplots
  geom_boxplot(width = rain_height, show.legend = FALSE, 
               outlier.shape = NA,
               position = position_nudge(x = -rain_height*2)) +
  coord_flip() +
  scale_fill_manual(values = c("#1a759f",
                               "#34a0a4",
                               "#52b69a",
                               "#99d98c",
                               "#d9ed92")) +
  scale_color_manual(values = c("#1a759f",
                                "#34a0a4",
                                "#52b69a",
                                "#99d98c",
                                "#d9ed92")) + 
  
  theme_pubr() +
  # stat_compare_means(method = "wilcox.test",
  #                    comparisons = list(c("TRUE", "FALSE")),
  #                    label = "p.format",
  #                    step.increase = 0.05) +
  theme(legend.position = "none") +
  
  labs(x = "Open intron", y = "Base Mean (log10)") +
  ggtitle("Overall base expression\nof open & closed introns")
all.samples.basemean.multi.genos





# What about the abs(log2FC) in response to heat for all samples

# Merge this with all genes in peaks (in at least 2 out of 4 samples)
# ID genes in peaks in genes 
in.peaks.two.genos.base.mean<- heat.response.results$gene %in% geneID_2_outta_4 #  genes overlap
heat.response.results$two.genos <- in.peaks.two.genos.base.mean
sum(heat.response.results$two.genos) # 1800

# Test the normality of the log10(baseMean) 
# log10 transform the baseMean
heat.response.results$abs.l2FC <- abs(heat.response.results$log2FoldChange)

# Sort the absolute values for plotting
sorted.abs.l2FC.all.samples <- sort(heat.response.results$abs.l2FC)

# Create a Q-Q plot
qqnorm(sorted.abs.l2FC.all.samples)
qqline(sorted.abs.l2FC.all.samples)

# Kolmogorov-Smirnov test for normality (can be applied to dataset > 5000)
ks.test(sorted.abs.l2FC.all.samples, "pnorm", mean = mean(sorted.abs.l2FC.all.samples), sd = sd(sorted.abs.l2FC.all.samples)) # reject the null that the data is normally distributed


all.samples.abs.l2fc.raw <- ggplot(heat.response.results, aes(x = two.genos, y = abs.l2FC, fill = two.genos)) +
  # clouds
  introdataviz::geom_flat_violin(trim=FALSE,
                                 position = position_nudge(x = rain_height+.05)) +
  # rain
  geom_point(aes(color = two.genos), size = 2, alpha = 0.4,  show.legend = FALSE, 
             position = position_jitter(width = rain_height, height = 0)) +
  # boxplots
  geom_boxplot(width = rain_height, show.legend = FALSE, 
               outlier.shape = NA,
               position = position_nudge(x = -rain_height*2)) +
  coord_flip() +
  scale_fill_manual(values = c("#a8dadc",
                               "#457b9d")) +
  scale_color_manual(values = c("#a8dadc",
                                "#457b9d")) + 
  
  theme_pubr() +
  stat_compare_means(method = "wilcox.test",
                     comparisons = list(c("TRUE", "FALSE")),
                     label = "p.format",
                     step.increase = 0.05) +
  theme(legend.position = "none") +
  
  labs(x = "Open intron", y = "Log2FC (absolute value)") +
  ggtitle("Gene log2FC following heat\nwith open & closed introns\n(All Samples)")
all.samples.abs.l2fc.raw 

# Filter out rows with high adj p-value
hist(heat.response.results$padj)
padj.filtered.heat.response.results <- heat.response.results[heat.response.results$padj < 0.05, ]
hist(padj.filtered.heat.response.results$padj)

# Sort the absolute values for plotting
sorted.abs.l2FC.all.samples.pval.fil <- sort(padj.filtered.heat.response.results$abs.l2FC)

# Create a Q-Q plot
qqnorm(sorted.abs.l2FC.all.samples.pval.fil)
qqline(sorted.abs.l2FC.all.samples.pval.fil)

# Kolmogorov-Smirnov test for normality (can be applied to dataset > 5000)
ks.test(sorted.abs.l2FC.all.samples.pval.fil, "pnorm", mean = mean(sorted.abs.l2FC.all.samples.pval.fil), sd = sd(sorted.abs.l2FC.all.samples.pval.fil)) # reject the null that the data is normally distributed


all.samples.abs.l2fc <- ggplot(padj.filtered.heat.response.results, aes(x = two.genos, y = abs.l2FC, fill = two.genos)) +
  # clouds
  introdataviz::geom_flat_violin(trim=FALSE,
                                 position = position_nudge(x = rain_height+.05)) +
  # rain
  geom_point(aes(color = two.genos), size = 2, alpha = 0.4,  show.legend = FALSE, 
             position = position_jitter(width = rain_height, height = 0)) +
  # boxplots
  geom_boxplot(width = rain_height, show.legend = FALSE, 
               outlier.shape = NA,
               position = position_nudge(x = -rain_height*2)) +
  coord_flip() +
  scale_fill_manual(values = c("#dec0f1",
                               "#aea3b0")) +
  scale_color_manual(values = c("#dec0f1",
                                "#aea3b0")) + 
  
  theme_pubr() +
  stat_compare_means(method = "wilcox.test",
                     comparisons = list(c("TRUE", "FALSE")),
                     label = "p.format",
                     step.increase = 0.05) +
  theme(legend.position = "none") +
  
  labs(x = "Open intron", y = "Log2FC (absolute value)") +
  ggtitle("Gene log2FC following heat\nwith open & closed introns\n(All Samples)")
all.samples.abs.l2fc 


# BaseMean (log 10) of HRGs in All samples 


# Checking for normality
# Sort the absolute values for plotting
sorted.lob10bM <- sort(padj.filtered.heat.response.results$log10bM)

# Create a Q-Q plot
qqnorm(sorted.lob10bM)
qqline(sorted.lob10bM)

# Kolmogorov-Smirnov test for normality (can be applied to dataset > 5000)
ks.test(sorted.lob10bM , "pnorm", mean = mean(sorted.lob10bM), sd = sd(sorted.lob10bM)) # reject the null that the data is normally distributed



all.samples.baseMean.HRG <- ggplot(padj.filtered.heat.response.results, aes(x = two.genos, y = log10bM, fill = two.genos)) +
  # clouds
  introdataviz::geom_flat_violin(trim=FALSE,
                                 position = position_nudge(x = rain_height+.05)) +
  # rain
  geom_point(aes(color = two.genos), size = 2, alpha = 0.4,  show.legend = FALSE, 
             position = position_jitter(width = rain_height, height = 0)) +
  # boxplots
  geom_boxplot(width = rain_height, show.legend = FALSE, 
               outlier.shape = NA,
               position = position_nudge(x = -rain_height*2)) +
  coord_flip() +
  scale_fill_manual(values = c("#f7b267",
                               "#f27059")) +
  scale_color_manual(values = c("#f7b267",
                                "#f27059")) + 
  
  theme_pubr() +
  stat_compare_means(method = "wilcox.test",
                     comparisons = list(c("TRUE", "FALSE")),
                     label = "p.format",
                     step.increase = 0.05) +
  theme(legend.position = "none") +
  
  labs(x = "Open intron", y = "Base Mean (log10)") +
  ggtitle("Heat response gene base mean\nwith open & closed introns")
all.samples.baseMean.HRG 

# All samples log2 FC of heat response genes
# Sort the absolute values for plotting
sorted.abs.log2FoldChange.all.samples <- sort(abs(padj.filtered.heat.response.results$log2FoldChange))

# Create a Q-Q plot
qqnorm(sorted.abs.log2FoldChange.all.samples)
qqline(sorted.abs.log2FoldChange.all.samples)

# Shapiro-Wilk test for normality
#shapiro.test(sorted.abs.log2FoldChange.all.samples) # Reject the null that the data is normally distributed
ks.test(sorted.abs.log2FoldChange.all.samples , "pnorm", mean = mean(sorted.abs.log2FoldChange.all.samples), sd = sd(sorted.abs.log2FoldChange.all.samples)) # reject the null that the data is normally distributed



l2fc.plot.all.samples <- ggplot(padj.filtered.heat.response.results, aes(x = two.genos, y = abs(log2FoldChange), fill = two.genos )) +
  # clouds
  introdataviz::geom_flat_violin(trim=FALSE,
                                 position = position_nudge(x = rain_height+.05)) +
  # rain
  geom_point(aes(color = two.genos), size = 2, alpha = 0.4,  show.legend = FALSE, 
             position = position_jitter(width = rain_height, height = 0)) +
  # boxplots
  geom_boxplot(width = rain_height, show.legend = FALSE, 
               outlier.shape = NA,
               position = position_nudge(x = -rain_height*2)) +
  coord_flip() +
  scale_fill_manual(values = c("#f7b267",
                               "#f27059")) +
  scale_color_manual(values = c("#f7b267",
                                "#f27059")) + 
  
  theme_pubr() +
  stat_compare_means(method = "wilcox.test",
                     comparisons = list(c("TRUE", "FALSE")),
                     label = "p.format",
                     step.increase = 0.05) +
  theme(legend.position = "none") +
  labs(x = "Open intron", y = "Log2FC (absolute value)") +
  ggtitle("Heat response log2FC\nof open & closed introns")
l2fc.plot.all.samples


# Load A. millepora Heat response genes

## HRG with abs(log2FC) >= 2
# 
# l2fc.plot.all.samples <- ggplot(amillepora.HRGs, aes(x = two.genos, y = abs(log2FoldChange), fill = two.genos )) +
#   # clouds
#   introdataviz::geom_flat_violin(trim=FALSE,
#                                  position = position_nudge(x = rain_height+.05)) +
#   # rain
#   geom_point(aes(color = two.genos), size = 2, alpha = 0.4,  show.legend = FALSE, 
#              position = position_jitter(width = rain_height, height = 0)) +
#   # boxplots
#   geom_boxplot(width = rain_height, show.legend = FALSE, 
#                outlier.shape = NA,
#                position = position_nudge(x = -rain_height*2)) +
#   coord_flip() +
#   scale_fill_manual(values = c("#f7b267",
#                                "#f27059")) +
#   scale_color_manual(values = c("#f7b267",
#                                 "#f27059")) + 
#   
#   theme_pubr() +
#   stat_compare_means(method = "wilcox.test",
#                      comparisons = list(c("TRUE", "FALSE")),
#                      label = "p.format",
#                      step.increase = 0.05) +
#   theme(legend.position = "none") +
#   labs(x = "Accessible intron", y = "Log2FC (absolute value)") +
#   ggtitle("Heat-stress Log2FC\nof Accessible & Non-accessible introns")
# l2fc.plot.all.samples

# Patchwork
(all.samples.basemean + exp.variation.plot) /
  (all.samples.baseMean.HRG + l2fc.plot.all.samples)

all.samples.basemean /
  exp.variation.plot

all.samples.baseMean.HRG /
  l2fc.plot.all.samples
### Updating Fisher's Exact test for the HRGs from all the heat stress assays
# Select for heat response genes
# hs.results <- lfcShrink(dds,
#                         contrast = c("condition.stress.txt", "36", "24.5"),
#                         type = "ashr")
# 
# heat.response.results <- hs.results[complete.cases(hs.results),]
# heat.response.results$gene <- rownames(heat.response.results)
# 
# in.peaks.two.genos.base.mean<- heat.response.results$gene %in% geneID_2_outta_4 #  genes overlap
# heat.response.results$two.genos <- in.peaks.two.genos.base.mean
# sum(heat.response.results$two.genos) # 896
# 
# # Filter out rows with high adj p-value
# hist(heat.response.results$padj)
# padj.filtered.heat.response.results <- heat.response.results[heat.response.results$padj < 0.05, ]
# hist(padj.filtered.heat.response.results$padj)
# 
# # Filter for change in heat 
amillepora.HRGs <- subset(padj.filtered.heat.response.results,
                          abs(log2FoldChange) >= 2)

# All HRGs
all.HRG.set <- as.data.frame(amillepora.HRGs$gene)
colnames(all.HRG.set) <- "gene"

# Count the overlap (yes HRG/yes Peak)
yes.HRG.yes.Peak <- sum(all.HRG.set$gene %in% union.peaks$gene)
yes.HRG.yes.Peak #680

# Set up contingency table
total.genes <- nrow(total.gene.and.peak) #5872, total 
HRG.count <- nrow(all.HRG.set) #2129
accessible.intron.genes.count <- nrow(accessible.peaks) # 1208, peaks in at least 2 out 4 samples 
overlap.count <- yes.HRG.yes.Peak # 680

# Create a 2x2 contingency table
contingency.table <- matrix(c(overlap.count, HRG.count - overlap.count,
                              accessible.intron.genes.count - overlap.count,
                              total.genes - HRG.count - accessible.intron.genes.count + overlap.count),
                            nrow = 2, ncol = 2, byrow = TRUE)

# Perform Fisher's exact test
result <- fisher.test(contingency.table)

# Extract p-value
p.value <- result$p.value
p.value # small p-value

